File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.go
func findPivot(nums []int, n int) int {
    start, end := 0, n - 1

    // Find the index of the pivot element using binary search.
    for start < end {
        // Skip duplicates at the beginning of the array.
        for start < end && nums[start] == nums[start + 1] {
            start++
        }
        // Skip duplicates at the end of the array.
        for start < end && nums[end] == nums[end - 1] {
            end--
        }

        mid := start + (end - start) / 2
        if nums[mid] >= nums[end] {
            start = mid + 1
        } else {
            end = mid
        }
    }

    return end
}

func bs(start int, end int, nums []int, target int) bool {
    // Binary search within a specific range [start, end].
    for start <= end {
        mid := start + (end - start) / 2
        if nums[mid] == target {
            return true
        } else if nums[mid] < target {
            start = mid + 1
        } else {
            end = mid - 1
        }
    }
    return false
}

func search(nums []int, target int) bool {
    n := len(nums)
    // Find the index where the array is rotated (the pivot).
    pivot := findPivot(nums, n)

    // Search for the target in the left subarray (before pivot).
    if bs(0, pivot - 1, nums, target) {
        return true
    }

    // Search for the target in the right subarray (after pivot).
    return bs(pivot, n - 1, nums, target)
}

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] nums, int target) {
        int n = nums.length;
        int low = 0, high = n - 1;

        while (low <= high) {
            // Calculate the middle index to avoid potential overflow
            int mid = low + (high - low) / 2;

            if (nums[mid] == target) {
                return true;
            }
            if ((nums[low] == nums[mid]) && (nums[mid] == nums[high])) {
                low++;
                high--;
            }
            // Check if the left half is sorted
            else if (nums[low] <= nums[mid]) {
                // If target is within the sorted left half, adjust the high pointer to search
                // in the left half
                if (target >= nums[low] && target < nums[mid]) {
                    high = mid - 1;
                }
                // Otherwise, adjust the low pointer to search in the right half
                else {
                    low = mid + 1;
                }
            }

            // If the right half is sorted
            else if (nums[mid] <= nums[high]) {
                // If target is within the sorted right half, adjust the low pointer to search
                // in the right half
                if (target > nums[mid] && target <= nums[high]) {
                    low = mid + 1;
                }
                // Otherwise, adjust the high pointer to search in the left half
                else {
                    high = mid - 1;
                }
            }
        }

        return false;
    }
}